Draft notebook and example output xml (#445)
BuildingSync/schema: docs/notebooks/bsync_examples/Small-Office-Level-2.md

# Energy Auditing with BuildingSync

> Follow the instructions in the [README.md](../README.md) file to run in Jupyter Lab/Notebook.

Author: I updated the README it's linking to and added a few sentences to make it a bit more clear

    print(etree.tostring(element.toxml(), pretty_print=True).decode('utf-8'))
    
def bsync_dump(root_element, file="example1.xml"):
    """Write the element to the specified file"""
    doctype = '<?xml version="1.0" encoding="UTF-8"?>'
    as_etree = root_element.toxml()
    as_etree.set("xmlns", "http://buildingsync.net/schemas/bedes-auc/2019")
    #as_etree.set("xmlns:xsi", "http://www.w3.org/2001/XMLSchema-instance")
    #as_etree.set("xsi:schemaLocation", "http://buildingsync.net/schemas/bedes-auc/2019 https://raw.githubusercontent.com/BuildingSync/schema/v2.4.0/BuildingSync.xsd")
    # Have to manually set the version right now. Align release of bsyncpy to this version.
    as_etree.set("version", "2.4.0")  
    output = etree.tostring(as_etree, doctype=doctype, pretty_print=True)
    with open(file, 'wb+') as f:
        f.write(output)
        return True
```

# Starting the Audit

## Small Office Prototype Building
The small office prototype is a single story rectangular building. We will use Denver as the location.

We can summarize the high level features as:
- Total Floor Area: 5500sf
- WWR: Southern Wall ~25%, Other Walls ~20%

> - Climate Zone 5B

Reviewer: ```suggestion
- ASHRAE Climate Zone 5B
```

- `Systems` are defined under a `Facility` and capture the energy consuming / producing assets. Each can be linked back to an individual `Facility`, `Site`, `Building`, or `Section`.
- `Reports` are defined under a `Facility` and each `Report` would contain information about reporting type requirements. Most importantly, a `Report` contains a set of `Scenario` elements, where each `Scenario` would convey building performance for one of: current / historical data, target performance, benchmark performance, baseline modeled performance, or expected performance with measure(s) implemented.

```python
# just take this for what it is at the moment. the bsync package makes
# it easier to work with xml content. All of this is mainly syntactic
# sugar around the lxml package, with some type checking, attribute 
# assignment, etc.
root = bsync.BuildingSync()
facs = bsync.Facilities()
f1 = bsync.Facilities.Facility(ID="Facility-1")
sites = bsync.Sites()
s1 = bsync.Sites.Site(ID="Site-1")
bldgs = bsync.Buildings()

> b1 = bsync.Buildings.Building(ID="Building-Small-Office-Prototype")

Reviewer: do we need to call this level 2?

sections = bsync.Sections()
systems = bsync.Systems()
contacts = bsync.Contacts()
reports = bsync.Reports()
measures = bsync.Measures()
schedules = bsync.Schedules()
utilities = bsync.Utilities()

# We link the entire report to the building
r1 = bsync.Report(
    bsync.LinkedPremisesOrSystem(
        bsync.LinkedPremisesOrSystem.Building(
            bsync.LinkedBuildingID(IDref=b1["ID"])
        )
    ),

>     ID='Report-L2-Audit')

Reviewer: There is `ASHRAEAuditLevel` that can be set here too. The syntax would be something like (confirm that it works). Also, if it does work, then add to Level 1 as well, please. (Element is [here](https://buildingsync.net/dictionary/2.4.0/#59990)).

```
r1 = bsync.Report(
     bsync.ASHRAEAuditLevel("Level 2: Energy Survey and Analysis")
     bsync.LinkedPremisesOrSystem(
         bsync.LinkedPremisesOrSystem.Building(
             bsync.LinkedBuildingID(IDref=b1["ID"])
         )
     ),
     ID='Report-L2-Audit')
```

scenarios = bsync.Scenarios()

# We now combine these and pretty print the result for easy viewing of what we just made
root += facs
facs += f1

# Direct children of a facility
f1 += sites
f1 += systems
f1 += schedules
f1 += measures
f1 += reports
f1 += contacts

# Specific to a site
sites += s1
s1 += bldgs
bldgs += b1

# Direct children of a building
b1 += sections

# Specific to a report
reports += r1
r1 += scenarios
r1 += utilities

pretty_print(root)
```

#### 6.2.1.1 Building information

This section of Standard 211 asks for lots of detail. We walk through the components of this that get added _as child elements of the building_, and then discuss other information.

```python
# 6.2.1.1.a - name
b1 += bsync.PremisesName('Small Office Prototype in Denver')

# 6.1.1.1.m 
b1 += bsync.PremisesNotes("Here we record general problems / issues identified in a walkthrough survey.")

# 6.1.1.1.d - address
b1 += bsync.Address(
    bsync.StreetAddressDetail(
        bsync.Simplified(
            bsync.StreetAddress("Some address")
        )
    ),
    bsync.Address.State("CO"),
    bsync.City("Denver"),

>     bsync.PostalCode("80401")

Reviewer: 80401 isn't in denver, haha. That's golden...

b1 += bsync.YearOfConstruction(2000)
b1 += bsync.YearOccupied(2000)
b1 += bsync.RetrocommissioningDate(date(2020,3,1))
b1 += bsync.YearOfLastMajorRemodel(2000)
b1 += bsync.YearOfLastEnergyAudit(2010)

#pretty_print(root)
```

Continuing with the information required by 6.1.1.1, we define additional content that doesn't sit as direct child elements of the Building. This includes:
- Contacts. Key contacts (1 auditor and 1 owner) __must be linked__ back to the building, even though they are not in the Building subtree:
    - Contacts live: /BuildingSync/Facilities/Facility/Contacts/Contact
    - An auditor is linked to a building at a Report level (i.e. who performed the audit)
    - Reports live: /BuildingSync/Facilities/Facility/Reports/Report
    - An audit report, with the correct links to a Building, Measures, etc. should be sufficient to meet the Standard 211 requirements spelled out in Section 6. More on this later.
- Space function breakdown by space type
- Occupied hours and number of occupants

#### Contacts

```python
# 6.1.1.1.b - Owner and auditor contact info
c1 = bsync.Contact(

>     bsync.ContactName('The dude'),

Reviewer: We are mixing double quotes and single quotes, pick one. I prefer single quotes for most python strings.

Author: Based on our conversation, we will update the whole file to have double quotes as xml prefers that, but let's keep this PR in single quotes for now.

        bsync.PartialOperationPercentage(23.)
    ),
    bsync.ScheduleDetail(
        bsync.DayType("Weekday"),
        bsync.ScheduleCategory("Lighting"),
        bsync.DayStartTime(datetime.strptime("23:00:00", "%H:%M:%S").time()),
        bsync.DayEndTime(datetime.strptime("23:59:59", "%H:%M:%S").time()),
        bsync.PartialOperationPercentage(18.)
    ),
    bsync.ScheduleDetail(
        bsync.DayType("Weekend"),
        bsync.ScheduleCategory("Lighting"),
        bsync.DayStartTime(datetime.strptime("00:00:00", "%H:%M:%S").time()),
        bsync.DayEndTime(datetime.strptime("23:59:59", "%H:%M:%S").time()),
        bsync.PartialOperationPercentage(18.)
    ),
    bsync.ScheduleDetail(
        bsync.DayType("Holiday"),
        bsync.ScheduleCategory("Lighting"),
        bsync.DayStartTime(datetime.strptime("00:00:00", "%H:%M:%S").time()),
        bsync.DayEndTime(datetime.strptime("23:59:59", "%H:%M:%S").time()),
        bsync.PartialOperationPercentage(18.)
    )
)
schedule_light += bsync.LinkedPremises(
    bsync.LinkedPremises.Building(
        bsync.LinkedBuildingID(IDref=b1['ID'])
    )
)
# plug load schedule

> schedule_pl += bsync.ScheduleDetails(

Reviewer: this is great. I'd really like to have a helper function in `bsyncpy` to more rapidly generate these sections, something like 

def create_bsync_schedule_detail

where it takes in a tabular view of the data, then simple converts to the detail bsync. 

someday.

Reviewer 2: Makes sense

        bsync.DayType("Weekday"),
        bsync.ScheduleCategory("HVAC equipment"),
        bsync.DayStartTime(datetime.strptime("20:00:00", "%H:%M:%S").time()),
        bsync.DayEndTime(datetime.strptime("23:59:59", "%H:%M:%S").time()),
        bsync.PartialOperationPercentage(0.)
    ),
    bsync.ScheduleDetail(
        bsync.DayType("Weekend"),
        bsync.ScheduleCategory("HVAC equipment"),
        bsync.DayStartTime(datetime.strptime("00:00:00", "%H:%M:%S").time()),
        bsync.DayEndTime(datetime.strptime("23:59:59", "%H:%M:%S").time()),
        bsync.PartialOperationPercentage(0.)
    ),
    bsync.ScheduleDetail(
        bsync.DayType("Holiday"),
        bsync.ScheduleCategory("HVAC equipment"),
        bsync.DayStartTime(datetime.strptime("00:00:00", "%H:%M:%S").time()),
        bsync.DayEndTime(datetime.strptime("23:59:59", "%H:%M:%S").time()),
        bsync.PartialOperationPercentage(0.)
    )
)
schedule_hvac_new += bsync.LinkedPremises(
    bsync.LinkedPremises.Building(
        bsync.LinkedBuildingID(IDref=b1['ID'])
    )
)

schedules += schedule_occ
schedules += schedule_light
schedules += schedule_pl
schedules += schedule_hvac
schedules += schedule_hvac_new

> #pretty_print(schedules)

Reviewer: ```suggestion

```

```

#### 6.2.1.2 Envelope

Level 2 energy audits require information about the building envelop, including roof, wall, floor and fenestration. BuildingSync defines these components (construction, thermal properties, etc.) under the `Systems` element, and links corresponding elements back to the envelope specified under the `Section` elements. A few other information (e.g. gross areas) are added under the `Building` element above.
In addition, 6.2.1.2.e requires information of the building overall tightness, and this is covered under the `AirInfiltrationSystems` element.

```python
# 6.2.1.2.a roof
roofsys = bsync.RoofSystems()
roof = bsync.RoofSystem(
    bsync.RoofConstruction("Wood frame"),
    bsync.RoofUFactor(4.706),
    ID=f"Roof-1"
)
roofsys += roof

> # 6.2.1.2.b wall

Reviewer 2: ```suggestion
# 6.2.1.2.b Opaque walls (above ground)
```

```

#### HVAC, SHW/DHW, Lighting, Plugs 

Level 2 energy audits require information about the primary systems serving a specific section, with details such as types, capacities and conditions

```python
# 6.2.1.3 HVAC
# Here we specify HVAC system details
# and link them back to the Section of interest.
# According to prototype small office spec in ASHRAE901_OfficeSmall_STD2004
# the heating source is Air-source heat pump with gas furnace as back up
# and cooling source is Air-source heat pump
# and air distribution unit is Single zone, constant air volume air distribution, one unit per occupied thermal zone

hvac_systems = bsync.HVACSystems()
hvac_system = bsync.HVACSystem(ID="HVACSystem-1")
hvac_systems += hvac_system
# link hvac to section
hvac_system += bsync.LinkedPremises(
    bsync.LinkedPremises.Section(
        bsync.LinkedSectionID(
            bsync.LinkedScheduleIDs(
                bsync.LinkedScheduleID(IDref=schedule_hvac['ID'])
            ),
            IDref=section["ID"]
        )
    )
)

# heating and cooling system
hc_systems = bsync.HeatingAndCoolingSystems()
hc_systems += bsync.ZoningSystemType("Single zone")

# cooling system
css = bsync.CoolingSources()

> coolingcapacity = [6862.6, 6344.4, 4688.8, 5836.3, 5989.0]

Reviewer: ```suggestion
coolingcapacities = [6862.6, 6344.4, 4688.8, 5836.3, 5989.0]
```

> for each in range(1,6):

Reviewer: maybe just iterate over the cooling capacities?

```suggestion
for index, coolingcapacity in enumerate(coolingcapacities):
```

    cs = bsync.CoolingSource(
        bsync.CoolingSourceType(
            bsync.DX(
                bsync.DXSystemType("Packaged/unitary heat pump"),
                bsync.CompressorType("Reciprocating"),
                bsync.CompressorStaging("Single stage")
            )
        ),
        bsync.AnnualCoolingEfficiencyValue(2.61),
        bsync.AnnualCoolingEfficiencyUnits('COP'),

>         bsync.Capacity(coolingcapacity[each-1]),

Reviewer: ```suggestion
        bsync.Capacity(coolingcapacity),
```

        bsync.CapacityUnits('W'),
        bsync.CoolingSourceCondition('Good'),
        bsync.CoolingSource.Controls(
            bsync.CoolingSource.Controls.Control(
                bsync.CoolingSource.Controls.Control.OtherControlTechnology(
                    bsync.ControlSystemType(
                        bsync.Digital()
                    )
                )
            )
        ),
        bsync.YearInstalled(2000),
        ID=f"CoolingSource-{each}"
    )
    css += cs

hc_systems += css

# heating system
hss = bsync.HeatingSources()

> heatingcapacity = [7203.66, 6659.65, 4921.86, 6126.30, 6286.66]

Reviewer 2: ```suggestion
heatingcapacities = [7203.66, 6659.65, 4921.86, 6126.30, 6286.66]
```

for each in range(1,6):
    hs = bsync.HeatingSource(
        bsync.HeatingSourceType(
            bsync.HeatingSourceType.HeatPump(
                bsync.HeatPumpType("Packaged Unitary"),
                bsync.HeatPumpBackupSystemFuel("Natural gas"),
                bsync.HeatPumpBackupAFUE(0.0),

>                 bsync.CoolingSourceID(IDref=cs['ID'])

Reviewer 2: Where is HeatingSourceID defined?

            )
        ),
        bsync.AnnualHeatingEfficiencyValue(3.01),
        bsync.AnnualHeatingEfficiencyUnits('COP'),

>         bsync.InputCapacity(heatingcapacity[each-1]),

Reviewer 2: same as Nick mentioned above

all_ru += cooling_ru
all_ru += shw_ru
all_ru += pl_ru
# ResourceUses are child elements of a specific scenario:
cbms += all_ru
```

#### All TimeSeries Data
The following cell simply performs the following:
1. Create a simple function to generate monthly timeseries elements for each resource use
1. Add it to a `TimeSeriesData` parent element
1. Add this back to the previously defined current building measured scenario (i.e. Scenario-1)

```python
full_ts_data = bsync.TimeSeriesData()

# usage and peak data
monthly_elec = [5663.92, 5099.31, 5290.08, 4868.67, 5121.08, 6018.81, 5776.42, 6407.08, 5702.81, 5139.72, 5080.94, 5204.11]
monthly_ng = [1.83, 2.32, 0.45, 0.62, 0.04, 0.0, 0.0, 0.0, 0.0, 0.0, 0.47, 0.63]
monthly_elec_peak = [19.68, 19.11, 17.02, 19.53, 16.04, 19.94, 18.65, 18.57, 18.48, 16.81, 17.36, 18.68]
monthly_elec_lf = [0.373, 0.394, 0.390, 0.358, 0.374, 0.386, 0.369, 0.372, 0.373, 0.354, 0.380, 0.362]

# costs data
monthly_elec_cost = [448.20, 443.41, 436.22, 443.97, 431.66, 450.82, 444.74, 447.31, 443.74, 434.72, 436.57, 442.23]
monthly_ng_cost = [46.87, 47.68, 44.62, 44.89, 43.95, 43.88, 43.88, 43.88, 43.88, 43.88, 44.65, 44.91]

elec_ids = []
ng_ids = []

> daysinmonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

Author: Let me figure it out in next update on this!

bench_st += bench
```

### Target Scenario
Relevant Standard 211 Sections:
- 6.1.4

Since we used a PM score in the baseline, we will also use that for our target. Let's say we are shooting for a target score of 70.

```python
# define the target scenario in reference to the benchmark scenario
target_sc = bsync.Scenario(
    bsync.AllResourceTotals(
        bsync.AllResourceTotal(
            bsync.AllResourceTotal.SiteEnergyUse(190348.6),  
            bsync.SiteEnergyUseIntensity(34.6),
            bsync.EnergyCost(4846.04),
            bsync.EnergyCostIndex(0.881),
            ID="AllResourceTotal-Target"
        )
    ),
    bsync.LinkedPremises(
        bsync.LinkedPremises.Building(
            bsync.LinkedBuildingID(IDref=b1['ID'])
        )
    ),
    ID="Scenario-Target"
)
target_st = bsync.Scenario.ScenarioType()
target = bsync.Target(
    bsync.ReferenceCase(IDref=bench_sc["ID"]),
    bsync.AnnualSavingsSiteEnergy(39070.53),
    bsync.AnnualSavingsCost(995),
    bsync.ENERGYSTARScore(89.),
)

# 
scenarios += target_sc
target_sc += target_st
target_st += target
```

### Current Building Modeled Scenario
Relevant Standard 211 Sections:
- 6.1.5 & 6.1.6

> Although not explicitly called out in Standard 211, the current building modeled scenario is mostly implied as part of a Level 2 energy audit when doing more detailed savings estimates / calculations for potential measure(s) implementation. This is because when an energy / cost savings claim is made for a package of measures scenario, it needs to be _in reference_ to something, i.e. a current building modeled scenario (also often referred to as a baseline modeled scenario). The baseline modeled scenario should be interpreted as the expected performance of your building on an average or typical year. This is assuming the baseline modeled scenario is performed with TMY3 data, although they are likely first calibrated with AMY data.

Reviewer: ```suggestion
Although not explicitly called out in Standard 211, the current building modeled scenario is implied as part of a Level 2 energy audit when doing detailed savings estimates and calculations for potential energy conservation measures. This is because when an energy and cost savings claim is made for a package of measures scenario, it needs to be _in reference_ to something, i.e., a current building modeled scenario (also often referred to as a baseline modeled scenario). The baseline modeled scenario should be interpreted as the expected performance of your building on an average or typical year. This is assuming the baseline modeled scenario is performed with TMY data, although they are likely first calibrated with AMY data.
```

            bsync.PackageFirstCost(0.),
            bsync.MVCost(0.),
            bsync.OMCostAnnualSavings(123.),
            bsync.EquipmentDisposalAndSalvageCosts(0.),
            bsync.ProjectMarkup(123.),
            bsync.FundingFromIncentives(123.),
            bsync.FundingFromTaxCredits(123.),
            bsync.OtherFinancialIncentives(123),
            bsync.RecurringIncentives(123),
            bsync.InternalRateOfReturn(1.),
            bsync.SimplePayback(123.),
            ID="POM-HVAC-Schedule"
        )
    ),
    bsync.LinkedPremises(
        bsync.LinkedPremises.Building(
            bsync.LinkedBuildingID(IDref=b1['ID'])
        )
    ),
    ID="Scenario-POM-HVAC-Schedule"
)
```

```python
scenarios += pom_sc_1
scenarios += pom_sc_2
scenarios += pom_sc_3
```

# Validation

So did what we just went through actually work? Do we have the required information for a Level 1 Audit? Head over to the [use case validator](https://buildingsync.net/validator) to find out!

![Selection Tool](./img/UC-Selection.png)

Use the line below to write the file to disk

```python
bsync_dump(root, file="example-level2.xml")
```

>     True

Author: Generated from the dump function. Deleted.
